### PR TITLE
[FIX] mail,test_mail: message_route; filter emails with wrong domain

### PR DESCRIPTION
**Pull request by nboulif**
Steps to reproduce:

  - Install CRM and Helpdesk modules (for test purposes)
  - Set a custom alias domain (e.g. "mydomain.com")
  - Go to CRM > Configuration > Sales Teams
  - Check that a team has en email Alias (e.g. "info@mydomain.com")
  - Go to Helpdesk > Configuration > Helpdesk Teams
  - Check that a team has en email Alias (e.g. "support@mydomain.com")
  - Email your instance with the following `to` value:
    info@mydomain.com, support@test.com
    (notice second email does not match the DB alias domain)
  - Go to CRM : A task has been created
  - Go to Helpdesk : A ticket has been created

Issue:

  The ticket in Helpdesk should not have been created.

Cause:

  The message_route method does not check the domain of the email
  address before creating the routes.

Solution:

  If `mail.catchall.domain.allowed` system parameter is set, filter to
  only keep the emails address that match the allowed domains (including
  domain set in `mail.catchall.domain` system parameter).

  The value of `mail.catchall.domain.allowed` system parameter should
  be a comma separated list of domains. e.g. `example.com,example.org`

opw-3150972

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
I fear this is going to break notably email forwarding, and this is a big functional change in behavior. Changing the way we route emails without notice in stable is risky ^^ .

See notably https://github.com/odoo/odoo/pull/38459 and https://github.com/odoo/odoo/pull/41042 . To discuss wednesday maybe, I think we could try to introduce back an optional configuration parameter with an allowed list of mail domains, but standard behavior should not change (until we move forward with multicompany alias management).

**Comment by nboulif** [comment]
Hello @tde-banana-odoo , do you think the tests and the validation check are ok? Thanks in advance!

File: addons/mail/models/ir_config_parameter.py [hunk 1]
```diff
@@ -1,21 +1,35 @@
...
 
-from odoo import api, models
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
 
 
 class IrConfigParameter(models.Model):
     _inherit = 'ir.config_parameter'
 
+    def _check_mail_catchall_allowed_domains(self, value):
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_config_parameter.py:11]: :socks: would define it after the main methods (aka after create / write), as CRUD methods should generally be on top of the file :)

File: addons/mail/models/ir_config_parameter.py [hunk 2]
```diff
@@ -1,21 +1,35 @@
...
+        error_message = _('Invalid value for `mail.catchall.allowed_domains` (must be a comma seperated list of domains, e.g. example.com,example.org )')
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_config_parameter.py:12]: Maybe ``Value for `mail.catchall.allowed_domains` cannot be validated. It should be a comma separated list of domains e.g. example.com,example.org.`` (no parenthesis, and also there is a typo ^^ )

File: addons/mail/models/ir_config_parameter.py [hunk 3]
```diff
@@ -1,21 +1,35 @@
...
+        if not value:
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_config_parameter.py:13]: Well giving a void value should probably be supported as no restriction would say, to be less strict. That way we can do "no key or no value -> no validation".

File: addons/mail/models/ir_config_parameter.py [hunk 4]
```diff
@@ -1,21 +1,35 @@
...
+            raise ValidationError(error_message)
+        value = [domain.strip() for domain in value.split(',') if domain]
+        if not value:
+            raise ValidationError(error_message)
+        return ",".join(value)
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_config_parameter.py:18]: As you reformat, maybe ``_clean_and_check_mail_catchall_allowed_domains`` ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 5]
```diff
@@ -801,6 +801,86 @@ def test_message_route_alias_multiple_new(self):
         new_simple = self.env['mail.test.gateway'].search([('name', '=', 'Test Subject')])
         self.assertEqual(len(new_simple), 1, 'message_process: a new mail.test should have been created')
 
+    @mute_logger('odoo.addons.mail.models.mail_thread', 'odoo.models')
+    def test_message_route_allowed_domains(self):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:805]: ``test_message_route_alias_with_allowed_domains`` ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, thanks for the PR ! Some validation on my side :) 

Cheers !

**Comment by nboulif** [comment]
Hello @tde-banana-odoo , thanks for all the checks (and big sorry the mess), I think applied all your notes :+1:

File: addons/test_mail/tests/test_mail_gateway.py [hunk 6]
```diff
@@ -801,6 +801,73 @@ def test_message_route_alias_multiple_new(self):
         new_simple = self.env['mail.test.gateway'].search([('name', '=', 'Test Subject')])
         self.assertEqual(len(new_simple), 1, 'message_process: a new mail.test should have been created')
 
+    @mute_logger('odoo.addons.mail.models.mail_thread', 'odoo.models')
+    def test_mail_catchall_domain_allowed_validation(self):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:805]: :socks: in the same file, there is a class 'TestMailAlias' holding alias-specific tests, this one could be moved in that class :)  like 'test_alias_domain_allowed_validation' to matche current naming in that class

File: addons/mail/models/ir_config_parameter.py [hunk 7]
```diff
@@ -12,10 +13,24 @@ def create(self, vals_list):
...
                 vals['value'] = self.env['mail.alias']._clean_and_check_unique(vals.get('value'))
+            elif 'value' in vals and parameter.key == 'mail.catchall.domain.allowed' and vals['value'] != parameter.value:
+                vals['value'] = vals['value'] and self._clean_and_check_mail_catchall_allowed_domains(vals['value'])
         return super().write(vals)
+
+    def _clean_and_check_mail_catchall_allowed_domains(self, value):
+        """ The purpose of this system parameter is to avoid the creation
+        of records from incoming emails with a domain != alias_domain
+        but that have a pattern matching an internal mail.alias . """
+        value = [domain.strip() for domain in value.split(',') if domain]
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_config_parameter.py:32]: To match what we do in the mailgateway (where we lower all found domains in rcpt_tos_localparts), shouldn't we `lower()` the domains ? That way we always compare lowered versions of it ?

File: addons/mail/models/mail_thread.py [hunk 8]
```diff
@@ -888,6 +888,10 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
         if not isinstance(message, EmailMessage):
             raise TypeError('message must be an email.message.EmailMessage at this point')
         catchall_alias = self.env['ir.config_parameter'].sudo().get_param("mail.catchall.alias")
+        catchall_domain = self.env["ir.config_parameter"].sudo().get_param("mail.catchall.domain")
+        catchall_domains_allowed = self.env["ir.config_parameter"].sudo().get_param("mail.catchall.domain.allowed")
+        if catchall_domain and catchall_domains_allowed:
+            catchall_domains_allowed = catchall_domains_allowed.split(',') + [catchall_domain]
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:894]: We could also lower catchall_domain, to lower everything

File: addons/test_mail/tests/test_mail_gateway.py [hunk 9]
```diff
@@ -801,6 +801,73 @@ def test_message_route_alias_multiple_new(self):
...
+        """ Check the validation of `mail.catchall.domain.allowed` system parameter"""
+        for value in (',', ',,'):
+            with self.assertRaises(exceptions.ValidationError):
+                self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', value)
+
+        for value in ('', 'hello.com', 'hello.com,,', 'hello.com,bonjour.com', 'hello.com, bonjour.com'):
+            self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', value)
+            value = ",".join([domain.strip() for domain in value.split(',') if domain]) if value else False
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:813]: Note that it would probably be better to give expected value, otherwise you might have issues in this line that hide potential errors. Idea is generally to have something explicit like
```
for input, expected in [
    ('', ''),
    ('hello.com', 'hello.com'),
    ('hello.com, bonjour.com', 'hello.com,bonjour.com'),
    ('hello.COM, BONJOUR.com', 'hello.com,bonjour.com'),
]:
    ...
    self.assetEqual(parsed_input, expected)
```

File: addons/test_mail/tests/test_mail_gateway.py [hunk 10]
```diff
@@ -801,6 +801,73 @@ def test_message_route_alias_multiple_new(self):
...
         self.assertEqual(len(new_simple), 1, 'message_process: a new mail.test should have been created')
 
+    @mute_logger('odoo.addons.mail.models.mail_thread', 'odoo.models')
+    def test_mail_catchall_domain_allowed_validation(self):
+        """ Check the validation of `mail.catchall.domain.allowed` system parameter"""
+        for value in (',', ',,'):
+            with self.assertRaises(exceptions.ValidationError):
+                self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', value)
+
+        for value in ('', 'hello.com', 'hello.com,,', 'hello.com,bonjour.com', 'hello.com, bonjour.com'):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:811]: If we go for the lower, would add input with uppercase -> lowercase :)

File: addons/test_mail/tests/test_mail_gateway.py [hunk 11]
```diff
@@ -801,6 +801,73 @@ def test_message_route_alias_multiple_new(self):
...
+
+        # test@.. will cause the creation of new mail.test
+        new_alias_2 = self.env['mail.alias'].create({
+            'alias_name': 'test',
+            'alias_user_id': False,
+            'alias_model_id': self.env['ir.model']._get('mail.test.container').id,
+            'alias_contact': 'everyone',
+        })
+
+        for subject, val1, val2, domain, sys_param in [
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:836]: What are val1 and val2 ?

Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_gateway.py:836]: val1 -> gateway_created ? (and make a boolean) ?
val2 -> container_created ? (and make a boolean) ?
domain -> alias2_domain ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 12]
```diff
@@ -801,6 +801,73 @@ def test_message_route_alias_multiple_new(self):
...
+            # and with a domain not allowed
+            ('Test Subject 1', 1, 1, not_allowed_domain, ""),
+            # Test with 'mail.catchall.domain.allowed' set in system parameters
+            # and with a domain not allowed
+            ('Test Subject 2', 1, 0, not_allowed_domain, allowed_domain),
+            # Test with 'mail.catchall.domain.allowed' set in system parameters
+            # and with a domain allowed
+            ('Test Subject 3', 1, 1, allowed_domain, allowed_domain),
+        ]:
+            self.env['ir.config_parameter'].set_param('mail.catchall.domain.allowed', sys_param)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:847]: Note that you could use "with self.subTest(...):", so that if one test crashes, the other tests are launched. It allows to test all use cases without having to fix them one by one, usefull in refactorings :)  you may grep existing code for some examples (you may give parameters to subtest, like ``with self.subTest(alias2_domain=alias2_domain, ...):``, and those are notably displayed in the console, allowing to know what are the parameters of the subtest)

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks ! While relooking at the code (have to be extra careful :D ) noticed that we `lower` the rcpt_to. We could probably do the same when cleaning the allowed domain lists to normalize a bit what we compare, what do you think ?

Some extra lint/coding comments, because well, it free comments day today :) .

Cheers !

File: addons/mail/models/mail_thread.py (hunk 8 above)
Comment (nboulif) [reply addons/mail/models/mail_thread.py:894]: Indeed, but I notice a weird behavior: The field `value` of a system parameter is `required`, but no value is set by default for `mail.catchall.domain`. As soon as we change it, we cannot clean it except by removing the record. I guess it might be ok since we should normally check first if the param exists (with `get_param`)...

However, if someone modify it and put anything in it (even just a space), it will not be cleaned/sanitized.
Question: Should we do a validation of the `mail.catchall.domain` (like for `mail.catchall.domain.allowed`) or just do `.strip().lower()` (and then check if there is a value) when using it?

File: addons/test_mail/tests/test_mail_gateway.py (hunk 12 above)
Comment (nboulif) [reply addons/test_mail/tests/test_mail_gateway.py:847]: Sorry, when I refactored the loop with your previous comment, I forgot to add the `with self.subTest`, my bad, thanks :pray:

File: addons/mail/models/mail_thread.py (hunk 8 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:894]: indeed we generally don't check if it contains space, ... we just check we have a value. Would say that it can be left as it is, as anyway you copy-paste your email domain in it. A comma-separated list (like this new param) is more subject to issues as people could make mistakes (double comma, spaces, ...).

TL;DR: would just do .strip().lower() here as suggested and let it untouched anywhere else

File: addons/mail/models/ir_config_parameter.py [hunk 13]
```diff
@@ -12,10 +13,24 @@ def create(self, vals_list):
...
+                vals['value'] = vals['value'] and self._clean_and_check_mail_catchall_allowed_domains(vals['value'])
         return super().write(vals)
+
+    def _clean_and_check_mail_catchall_allowed_domains(self, value):
+        """ The purpose of this system parameter is to avoid the creation
+        of records from incoming emails with a domain != alias_domain
+        but that have a pattern matching an internal mail.alias . """
+        value = [domain.strip().lower() for domain in value.split(',') if domain.strip().lower()]
+        if not value:
+            raise ValidationError(_("""Value for `mail.catchall.domain.allowed` cannot be validated.
```
Comment (tde-banana-odoo) [inline addons/mail/models/ir_config_parameter.py:34]: Do we need the line return ? When using triple quotes, line breaks are kept ...

File: addons/mail/models/mail_thread.py [hunk 14]
```diff
@@ -888,6 +888,10 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
         if not isinstance(message, EmailMessage):
             raise TypeError('message must be an email.message.EmailMessage at this point')
         catchall_alias = self.env['ir.config_parameter'].sudo().get_param("mail.catchall.alias")
+        catchall_domain = self.env["ir.config_parameter"].sudo().get_param("mail.catchall.domain", "").strip().lower()
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:891]: 'catchall_domain_lowered' ?

File: addons/mail/models/mail_thread.py [hunk 15]
```diff
@@ -917,10 +921,11 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
         ]
         # Delivered-To is a safe bet in most modern MTAs, but we have to fallback on To + Cc values
         # for all the odd MTAs out there, as there is no standard header for the envelope's `rcpt_to` value.
-        rcpt_tos_localparts = [
-            e.split('@')[0].lower()
-            for e in tools.email_split(message_dict['recipients'])
-        ]
+        rcpt_tos_localparts = []
+        for e in tools.email_split(message_dict['recipients']):
+            to = e.split('@')
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:926]: Maybe ``to_local_lowered, to_domain_lowered  = e.split('@', maxsplit=1)`` ? That way we tackle several potential issues: compare lower string, be sure that we split only once (consider all other '@' to be part of the domain) ?

File: addons/mail/models/mail_thread.py [hunk 16]
```diff
@@ -917,10 +921,11 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
         ]
         # Delivered-To is a safe bet in most modern MTAs, but we have to fallback on To + Cc values
         # for all the odd MTAs out there, as there is no standard header for the envelope's `rcpt_to` value.
-        rcpt_tos_localparts = [
-            e.split('@')[0].lower()
-            for e in tools.email_split(message_dict['recipients'])
-        ]
+        rcpt_tos_localparts = []
+        for e in tools.email_split(message_dict['recipients']):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:925]: e -> email_to ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 17]
```diff
@@ -801,6 +818,66 @@ def test_message_route_alias_multiple_new(self):
...
+                    subject=subject,
+                    target_model=self.alias.alias_model_id.model
+                )
+
+                res_alias_1 = MailTestGatewayModel.search([('name', '=', subject)])
+                res_alias_2 = MailTestContainerModel.search([('name', '=', subject)])
+                self.assertEqual(
+                    bool(res_alias_1), gateway_created,
+                    'message_process (%s): a new mail.test.gateway should%s have been created' %
+                        (subject, '' if gateway_created else "n't")
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:873]: haha, the "n't" stuff kept me looking at the screen for a few seconds :D ok, if you don't want to rewrite the message, maybe set the whole word in the variable (should / should not). Have to say I wondered if this was some magic :D

File: addons/test_mail/tests/test_mail_gateway.py [hunk 18]
```diff
@@ -236,6 +236,23 @@ def _test_alias_mixin_copy(self, user, alias_name, alias_bounced_content):
         record_copy = record.copy()
         self.assertEqual(record_copy.alias_bounced_content, alias_bounced_content)
 
+    def test_alias_domain_allowed_validation(self):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:239]: :socks: beware alphabetical ordering in a given class

**Review by tde-banana-odoo — COMMENTED** [review]
Keep faith, we will close this one someday :D

**Review by tde-banana-odoo — APPROVED** [review]

File: addons/mail/models/ir_config_parameter.py [hunk 19]
```diff
@@ -12,10 +13,24 @@ def create(self, vals_list):
...
                 vals['value'] = self.env['mail.alias']._clean_and_check_unique(vals.get('value'))
+            elif 'value' in vals and parameter.key == 'mail.catchall.domain.allowed' and vals['value'] != parameter.value:
+                vals['value'] = vals['value'] and self._clean_and_check_mail_catchall_allowed_domains(vals['value'])
         return super().write(vals)
+
+    def _clean_and_check_mail_catchall_allowed_domains(self, value):
+        """ The purpose of this system parameter is to avoid the creation
+        of records from incoming emails with a domain != alias_domain
+        but that have a pattern matching an internal mail.alias . """
+        value = [domain.strip().lower() for domain in value.split(',') if domain.strip().lower()]
```
Comment (luisg123v) [inline addons/mail/models/ir_config_parameter.py:32]: No need to lower when checking if value exists

```suggestion
        value = [domain.strip().lower() for domain in value.split(',') if domain.strip()]
```

**Review by luisg123v — CHANGES_REQUESTED** [review]
Typo in commit description:

"a comma seperated list" -> separated

**Comment by luisg123v** [comment]
FYI this also closes #69782

File: addons/mail/models/ir_config_parameter.py (hunk 19 above)
Comment (nboulif) [reply addons/mail/models/ir_config_parameter.py]: Indeed, thanks :+1:

File: addons/mail/models/mail_thread.py (hunk 16 above)
Comment (nboulif) [reply addons/mail/models/mail_thread.py]: Struggled with one test to understand why breaking and not every time... `email_to` was already used above :man_facepalming: using `recipient` instead

**Comment by yajo** [comment]
IIUC this new ICP is the same as `mail.default.from_filter`, right? Wouldn't it make more sense to backport https://github.com/odoo/odoo/pull/80943?

Of course this one supports more than 1 domain, which is a good feature that could be then forward-ported to other versions.

**Comment by yajo** [comment]
Ah, forget my last comment. I got confused. This PR is about incoming mail, whereas the from filter is for outgoing.

It seems to me like this should be a field in the fetchmail server model though, just like the from filter is for outgoing mail servers in v16. Maybe you could do that for master?

**Comment by tde-banana-odoo** [comment]
@yajo For master I still hope to land with the multi-company alias support, which will replace the configuration parameter completely normally.

**Comment by yajo** [comment]
Me too, but I feel that'll never happen... 🙄

**Comment by tde-banana-odoo** [comment]
> Me too, but I feel that'll never happen... roll_eyes

There is always hope.

**Review by rafaelbn — APPROVED** [review]
Thank you! @nboulif @tde-banana-odoo ! ❤️ 

This solves a good pack of OPWs!

👍🏼

**Comment by tde-banana-odoo** [comment]
@robodoo r+
